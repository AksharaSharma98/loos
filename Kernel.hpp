/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/





#if !defined(KERNEL_HPP)
#define KERNEL_HPP

#include <vector>

#include "Atom.hpp"

#include "KernelValue.hpp"
#include "KernelStack.hpp"
#include "KernelActions.hpp"

using namespace std;

namespace loos {

  //!The Kernel (virtual machine) for compiling and executing user-defined atom selections

  class Kernel {
    vector<Action*> actions;    //! Commands
    ValueStack val_stack;       //! The data stack...

  public:
    
    //! Destroy all of the stored commands...
    ~Kernel() {
      vector<Action*>::iterator i;
      for (i = actions.begin(); i != actions.end(); i++)
	delete (*i);
    }

    //! Add a command, setting the data-stack pointer...
    void push(Action *act) {
      act->setStack(&val_stack);
      actions.push_back(act); 
    }

    void pop(void) { actions.pop_back(); }

    //! Execute the stored commands for a specific atom.
    //! This does not clear the command stack/list
    void execute(pAtom pa = pAtom()) {

      vector<Action*>::iterator i;
      for (i=actions.begin(); i != actions.end(); i++) {
	(*i)->setAtom(pa);
	(*i)->execute();
      }
    }

    
    void clearActions(void) { actions.clear(); }

    ValueStack& stack(void) { return(val_stack); }

    friend ostream& operator<<(ostream& os, const Kernel& k) {
      vector<Action*>::const_iterator i;

      os << "Commands:\n";
      for (i=k.actions.begin(); i != k.actions.end(); i++)
	os << (*i)->name() << endl;
      os << endl;
      return(os);
    }

  };

};


#endif
