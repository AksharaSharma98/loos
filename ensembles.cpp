/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/





#include <ensembles.hpp>
#include <XForm.hpp>
#include <AtomicGroup.hpp>
#include <Trajectory.hpp>

namespace loos {

  // Assume all groups are already sorted or matched...

  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble) {
    AtomicGroup avg = ensemble[0].copy();

    // First, zap our coords...
    int n = avg.size();
    int i;
    for (i=0; i<n; i++)
      avg[i]->coords() = GCoord(0.0, 0.0, 0.0);

    // Now, accumulate...
    std::vector<AtomicGroup>::const_iterator j;
    for (j = ensemble.begin(); j != ensemble.end(); j++) {
      for (i = 0; i<n; i++)
        avg[i]->coords() += (*j)[i]->coords();
    }

    for (i=0; i<n; i++)
      avg[i]->coords() /= ensemble.size();

    return(avg);
  }



  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble, const std::vector<XForm>& xforms) {
    if (xforms.size() != ensemble.size())
      throw(LOOSError("Transforms do not match the passed ensemble in loos::averageStructure()"));

    AtomicGroup avg = ensemble[0].copy();

    // First, zap our coords...
    int n = avg.size();
    int i;
    for (i=0; i<n; i++)
      avg[i]->coords() = GCoord(0.0, 0.0, 0.0);

    // Now, accumulate...
    for (uint j=0; j<ensemble.size(); ++j) {
      AtomicGroup frame = ensemble[j].copy();
      frame.applyTransform(xforms[j]);
      for (i = 0; i<n; i++)
        avg[i]->coords() += frame[i]->coords();
    }
    
    for (i=0; i<n; i++)
      avg[i]->coords() /= ensemble.size();

    return(avg);
  }




  AtomicGroup averageStructure(const AtomicGroup& g, const std::vector<XForm>& xforms, pTraj& traj, std::vector<uint>& frame_indices) {
    AtomicGroup avg = g.copy();
    AtomicGroup frame = g.copy();
    int n = avg.size();
    for (int i=0; i<n; i++)
      avg[i]->coords() = GCoord(0.0, 0.0, 0.0);
    
    uint tn = traj->nframes();
    uint fn = frame_indices.size();
    
    // Safety check...
    if (fn != xforms.size())
      throw(std::runtime_error("Mismatch in number of frames in the trajectory requested and passed transforms for loos::averageStructure()"));
    
    for (uint j=0; j<fn; ++j) {
      if (frame_indices[j] >= tn)
        throw(std::runtime_error("Frame index exceeds trajectory size"));
      
      traj->readFrame(frame_indices[j]);
      traj->updateGroupCoords(frame);
      frame.applyTransform(xforms[j]);
      for (int i=0; i<n; i++)
        avg[i]->coords() += frame[i]->coords();
    }
    
    for (int i=0; i<n; i++)
      avg[i]->coords() /= fn;
    
    return(avg);
  }
  
  
  
  AtomicGroup averageStructure(const AtomicGroup& g, const std::vector<XForm>& xforms, pTraj& traj) {
    uint nx = xforms.size();
    uint nt = traj->nframes();
    
    if (nt != nx)
      throw(std::runtime_error("Mismatch in number of frames in the trajectory and passed transforms"));
    std::vector<uint> frame_indices;
    for (uint i=0; i<nt; ++i)
      frame_indices.push_back(i);

    return(averageStructure(g, xforms, traj, frame_indices));
  }





  boost::tuple<std::vector<XForm>,greal,int> iterativeAlignment(std::vector<AtomicGroup>& ensemble,
                                                                greal threshold, int maxiter) {
    int iter = 0;
    int n = ensemble.size();
    greal rms;
    std::vector<XForm> xforms(n);
    AtomicGroup avg;

    // Start by aligning against the first structure in the ensemble
    AtomicGroup target = ensemble[0].copy();

    target.centerAtOrigin();

    do {
      for (int i = 0; i<n; i++) {
        GMatrix M = ensemble[i].alignOnto(target);
        xforms[i].premult(M);
      }

      avg = averageStructure(ensemble);
      rms = avg.rmsd(target);
      target = avg;
      iter++;
    } while (rms > threshold && iter <= maxiter );
    
    boost::tuple<std::vector<XForm>, greal, int> res(xforms, rms, iter);
    return(res);
  }



  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj, std::vector<uint>& frame_indices,
                                                                  greal threshold, int maxiter) {
    std::vector<AtomicGroup> frames;

    uint nf = frame_indices.size();
    uint nt = traj->nframes();

    for (uint i=0; i<nf; ++i) {
      if (frame_indices[i] >= nt)
        throw(std::runtime_error("Frame index exceeds trajectory size"));

      traj->readFrame(frame_indices[i]);
      AtomicGroup frame = g.copy();
      traj->updateGroupCoords(frame);
      frames.push_back(frame);
    }

    boost::tuple<std::vector<XForm>, greal, int> result = iterativeAlignment(frames, threshold, maxiter);
    return(result);
  }


  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj,
                                                                  greal threshold, int maxiter) {

    std::vector<uint> frame_indices;
    uint nt = traj->nframes();
    for (uint i=0; i<nt; ++i)
      frame_indices.push_back(i);
    return(iterativeAlignment(g, traj, frame_indices, threshold, maxiter));
  }



  void applyTransforms(std::vector<AtomicGroup>& ensemble, std::vector<XForm>& xforms) {
    uint n = ensemble.size();
    if (n != xforms.size())
      throw(std::runtime_error("Mismatch in the size of the ensemble and the transformations"));

    for (uint i=0; i<n; ++i)
      ensemble[i].applyTransform(xforms[i]);
  }



  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory) {
    AtomicGroup clone = model.copy();
    
    while (trajectory->readFrame()) {
      trajectory->updateGroupCoords(clone);
      AtomicGroup frame = clone.copy();
      ensemble.push_back(frame);
    }
  }


  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory, std::vector<uint>& frames) {
    AtomicGroup clone = model.copy();
    
    std::vector<uint>::iterator i;
    for (i = frames.begin(); i != frames.end(); ++i) {
      if (*i >= trajectory->nframes())
        throw(std::runtime_error("Frame index exceeds trajectory size in readTrajectory()"));
      trajectory->readFrame(*i);
      trajectory->updateGroupCoords(clone);
      AtomicGroup frame = clone.copy();
      ensemble.push_back(frame);
    }
  }



  RealMatrix extractCoords(std::vector<AtomicGroup>& ensemble) {
    uint n = ensemble.size();
    uint m = ensemble[0].size();
    RealMatrix M(3*m, n);

    for (uint i=0; i<n; ++i)
      for (uint j=0; j<m; ++j) {
        M(3*j, i) = ensemble[i][j]->coords().x();
        M(3*j+1, i) = ensemble[i][j]->coords().y();
        M(3*j+2, i) = ensemble[i][j]->coords().z();
      }

    return(M);
  }


  RealMatrix extractCoords(std::vector<AtomicGroup>& ensemble, std::vector<XForm>& xforms) {
    uint n = ensemble.size();

    if (n != xforms.size())
      throw(std::runtime_error("Mismatch between the size of the ensemble and the transformations"));

    uint m = ensemble[0].size();
    RealMatrix M(3*m, n);

    for (uint i=0; i<n; ++i) {
      GMatrix W = xforms[i].current();

      for (uint j=0; j<m; ++j) {
        GCoord c = W * ensemble[i][j]->coords();
        M(3*j, i) = c.x();
        M(3*j+1, i) = c.y();
        M(3*j+2, i) = c.z();
      }
    }

    return(M);
  }


  void subtractAverage(RealMatrix& M) {
    uint m = M.rows();
    uint n = M.cols();
    double *avg = new double[m];
    
    for (uint j=0; j<m; ++j)
      avg[j] = 0.0;

    for (uint i=0; i<n; ++i)
      for (uint j=0; j<m; ++j)
        avg[j] += M(j, i);

    for (uint j=0; j<m; ++j)
      avg[j] /= n;

    for (uint i=0; i<n; ++i)
      for (uint j=0; j<m; ++j)
        M(j,i) -= avg[j];
  }


  boost::tuple<RealMatrix, RealMatrix, RealMatrix> svd(std::vector<AtomicGroup>& ensemble, bool align) {

    RealMatrix M;
    if (align) {
      boost::tuple<std::vector<XForm>, greal, int> res = iterativeAlignment(ensemble);
      M = extractCoords(ensemble, boost::get<0>(res));
    } else
      M = extractCoords(ensemble);

    subtractAverage(M);
    boost::tuple<RealMatrix, RealMatrix, RealMatrix> res = Math::svd(M);
    return(res);
  }



}
