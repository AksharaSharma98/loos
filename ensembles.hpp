/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/



#if !defined(ENSEMBLES_HPP)
#define ENSEMBLES_HPP

#include <vector>
#include <XForm.hpp>

#include <loos_defs.hpp>

namespace loos {
  //! Compute the average structure of a set of AtomicGroup objects
  AtomicGroup averageStructure(const vector<AtomicGroup>& ensemble);

  //! Compute the average structure by reading through a trajectory
  AtomicGroup averageStructure(const AtomicGroup&, const vector<XForm>&, pTraj);

  //! Compute an iterative superposition (a la Alan)
  boost::tuple<vector<XForm>, greal, int> iterativeAlignment(vector<AtomicGroup>& ensemble, greal threshold, int maxiter=1000);

  //! Compute an iterative superposition by reading in frames from the Trajectory.
  /*!
   * This function will internally cache an AtomicGroup copy for each
   * frame of the trajectory.  This could chew up a lot of memory, but
   * we make the assumption that you will usually be aligning against
   * a fairly small subset of each frame...
   */
  boost::tuple<vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj, greal threshold, int maxiter=1000);
};



#endif
