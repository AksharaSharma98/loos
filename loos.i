%module loos


%feature("autodoc", "1");

%include <std_string.i>
%include <std_vector.i>
%include <boost_shared_ptr.i>

typedef unsigned int    uint;
typedef unsigned long     ulong;

namespace loos {
  typedef double greal;
  typedef long gint;

  typedef float dcd_real;
  typedef double dcd_double;

  typedef Coord<double> GCoord;
}

%include "Coord.i"
%include "Atom.i"
%include "Matrix44.i"
%include "pdb_remarks.i"
%include "XForm.i"
%include "AtomicGroup.i"
%include "Trajectory.i"
%include "utils.i"
%include "cryst.i"
%include "pdb.i"
%include "sfactories.i"
%include "dcdwriter.i"
%include "ensembles.i"
%include "Geometry.i"
%include "TimeSeries.i"
%include "HBondDetector.i"
%include "exceptions.i"
