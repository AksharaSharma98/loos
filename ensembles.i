/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/



%header %{
#include <ensembles.hpp>
%}

namespace loos {

  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble);

  AtomicGroup averageStructure(const std::vector<AtomicGroup>& ensemble, const std::vector<XForm>& xforms);

  AtomicGroup averageStructure(const AtomicGroup&, const std::vector<XForm>&, pTraj& traj, const std::vector<uint>& indices);

  AtomicGroup averageStructure(const AtomicGroup&, const std::vector<XForm>&, pTraj& traj);

  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(std::vector<AtomicGroup>& ensemble, greal threshold = 1e-6, int maxiter=1000);

  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj, const std::vector<uint>& frame_indices, greal threshold = 1e-6, int maxiter=1000);

  boost::tuple<std::vector<XForm>, greal, int> iterativeAlignment(const AtomicGroup& g, pTraj& traj, greal threshold = 1e-6, int maxiter=1000);


  void applyTransforms(std::vector<AtomicGroup>& ensemble, std::vector<XForm>& xforms);

  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory);
  void readTrajectory(std::vector<AtomicGroup>& ensemble, const AtomicGroup& model, pTraj trajectory, std::vector<uint>& frames);
};

