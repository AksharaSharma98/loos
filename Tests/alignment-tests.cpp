/*
  alignment-tests.cpp
  (c) 2008 Tod D. Romo


  Grossfield Lab
  Department of Biochemistry and Biophysics
  University of Rochester Medical School

  Unit testing for alignment routines...

*/

#include <iostream>
#include <iomanip>
#include <boost/random.hpp>
#include <ctime>


#include <loos.hpp>
#include <pdb.hpp>
#include <Kernel.hpp>
#include <Parser.hpp>
#include <Selectors.hpp>
#include <ensembles.hpp>

// Configuration options...
const unsigned int iter_tests = 1000;
const double iter_perturbation = 2.0;
const double iter_rmsd_thresh = 2.0;
const double iter_final_rmsd_thresh = 1e-3;

const unsigned int single_tests = 1000;
const double single_rmsd_thresh = 1e-10;
const static bool show_results = false;



int main(int argc, char *argv[]) {
  unsigned int i;

  if (argc != 2) {
    cerr << "Usage- " << argv[0] << " pdbfile\n";
    exit(-1);
  }
  
  // Suppress for easy diffs...
  //cout << invocationHeader(argc, argv) << endl;

  loos::base_generator_type& rng = loos::rng_singleton();

  // Uncommont the follong to seed the suite-wide RNG
  //loos::randomSeedRNG();


  PDB pdb(argv[1]);

  CAlphaSelector casel;
  AtomicGroup cas = pdb.select(casel);
  cout << "Found " << cas.size() << " CAlphas.\n";

  AtomicGroup casb = cas.copy();
  casb.perturbCoords(1.0);
  greal rmsd = cas.rmsd(casb);
  cout << "RMSD test = " << rmsd << endl;
  XForm W;
  W.rotate('Y', 35.0);
  casb.applyTransform(W);
  rmsd = cas.rmsd(casb);
  cout << "Rotated rmsd = " << rmsd << endl;

  GMatrix M = casb.alignOnto(cas);
  cout << "Aligned rmsd = " << cas.rmsd(casb) << endl;
  cout << M << endl;


  boost::uniform_real<> anglemap(-180.0, 180.0);
  boost::variate_generator<loos::base_generator_type&, boost::uniform_real<> > angles(rng, anglemap);

  boost::uniform_real<> transmap(-20.0, 20.0);
  boost::variate_generator<loos::base_generator_type&, boost::uniform_real<> > translations(rng, transmap);


  // -------------------------------------------------------------------------------

  cout << "*** Single Alignment Tests ***\n";

  int warnings = 0;
  for (i=0; i<single_tests; i++) {
    AtomicGroup casr = cas.copy();
    W.identity();
    W.translate(translations(), translations(), translations());
    W.rotate('x', angles());
    W.rotate('y', angles());
    W.rotate('z', angles());
    casr.applyTransform(W);
    greal pre_rmsd = cas.rmsd(casr);
    casr.alignOnto(cas);
    double rmsd = cas.rmsd(casr);
    if (rmsd >= single_rmsd_thresh) {
      cout << "WARNING - Possible mis-alignment - pre = " << pre_rmsd << ", post = " << rmsd << endl;
      ++warnings;
    }
  }

  if (warnings > 0)
    cout << "*** There were " << warnings << " possible errors detected.\n";
  else
    cout << "All tests passed (threshold = " << single_rmsd_thresh << ")\n";



  // -------------------------------------------------------------------------------
  // Now run iteratative superpositon tests...

  vector<AtomicGroup> mols;
  for (i=0; i<iter_tests; i++) {
    AtomicGroup subgroup = cas.copy();
    GCoord t(translations(), translations(), translations());
    
    subgroup.perturbCoords(iter_perturbation);

    W.identity();
    W.translate(t);
    W.rotate('z', angles());
    W.rotate('y', angles());
    W.rotate('x', angles());
	     
    subgroup.applyTransform(W);
    mols.push_back(subgroup);
  }

  AtomicGroup avg = averageStructure(mols);
  if (show_results) {
    cout << "Pre-aligned rmsds:\n";
    for (i=0; i<iter_tests; i++)
      cout << "\t" << i << "\t" << avg.rmsd(mols[i]) << endl;
  }


  greal final_rmsd = loos::iterativeAlignment(mols, 2.0);
  if (show_results)
    cout << "Final alignment rmsd to avg struct = " << final_rmsd << endl;

  if (final_rmsd >= iter_final_rmsd_thresh)
    cout << "WARNING - final rmsd of " << final_rmsd << " is above threshold.\n";

  warnings = 0;
  avg = averageStructure(mols);
  for (i=0; i<iter_tests; i++) {
    if (show_results)
      cout << "\t" << i << "\t" << avg.rmsd(mols[i]) << endl;
    double irmsd = avg.rmsd(mols[i]);
    if (irmsd >= iter_rmsd_thresh) {
      ++warnings;
      cout << "WARNING - possible iterative failure at " << i << " with rmsd of " << irmsd << endl;
    }
  }

  if (warnings > 0)
    cout << "*** There were " << warnings << " possible errors detected.\n";
  else
    cout << "All tests passed (threshold = " << iter_final_rmsd_thresh << " : " << iter_rmsd_thresh << ")\n";
}
