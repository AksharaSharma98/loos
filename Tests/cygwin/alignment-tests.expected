

-=-=-=-=-=-=-=- MACROCYCLE #0 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.3013
Aligned rmsd = 0.975447
[0.829865 -0.00333348 -0.557954 -0.816296;-0.00911085 0.999768 -0.019524 0.912026;0.557889 0.0212857 0.829642 -0.858013;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 159 with rmsd of 2.00005
WARNING - possible iterative failure at 280 with rmsd of 2.00902
*** There were 2 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.0491
Aligned rmsd = 0.982821
[0.809957 0.0109384 -0.586388 -0.743293;-0.0177038 0.999826 -0.0058031 -0.627915;0.586222 0.0150816 0.81001 0.178563;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 49 with rmsd of 2.00732
*** There were 1 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.3713
Aligned rmsd = 0.988222
[0.816016 0.00509634 -0.578006 -0.231732;-0.00831081 0.999961 -0.00291626 -0.342981;0.577969 0.00718342 0.816027 -0.134515;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 290 with rmsd of 2.00056
WARNING - possible iterative failure at 416 with rmsd of 2.00038
WARNING - possible iterative failure at 684 with rmsd of 2.00027
WARNING - possible iterative failure at 776 with rmsd of 2.0014
WARNING - possible iterative failure at 804 with rmsd of 2.00125
WARNING - possible iterative failure at 896 with rmsd of 2.00055
*** There were 6 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #1 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.3393
Aligned rmsd = 0.979355
[0.816127 -0.0128346 -0.57773 1.08596;0.00799538 0.999908 -0.010919 -0.222239;0.577818 0.0042921 0.816155 -0.409091;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
All tests passed (threshold = 0.01 : 2)
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.3215
Aligned rmsd = 0.948189
[0.773359 0.0478944 -0.632157 -3.0518;-0.0332916 0.998834 0.0349474 -0.885428;0.633094 -0.00598136 0.774052 1.98979;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 237 with rmsd of 2.00157
*** There were 1 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.3555
Aligned rmsd = 0.997971
[0.817856 -0.0113101 -0.575312 0.271393;0.00812371 0.999934 -0.00810927 0.197811;0.575366 0.00195855 0.817894 -0.0355369;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 119 with rmsd of 2.00136
WARNING - possible iterative failure at 337 with rmsd of 2.00373
WARNING - possible iterative failure at 581 with rmsd of 2.00096
WARNING - possible iterative failure at 676 with rmsd of 2.00002
WARNING - possible iterative failure at 677 with rmsd of 2.00089
WARNING - possible iterative failure at 691 with rmsd of 2.00206
WARNING - possible iterative failure at 723 with rmsd of 2.00052
*** There were 7 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #2 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.2115
Aligned rmsd = 0.979267
[0.816411 -0.00886837 -0.577404 0.885106;0.0152327 0.999865 0.0061811 -1.13565;0.577271 -0.0138418 0.816435 0.837822;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
All tests passed (threshold = 0.01 : 2)
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.3784
Aligned rmsd = 0.949022
[0.835912 0.0702561 -0.544349 -0.747798;-0.031939 0.99632 0.0795434 -0.769868;0.547934 -0.0491053 0.835079 0.384109;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 28 with rmsd of 2.00307
WARNING - possible iterative failure at 483 with rmsd of 2.00464
WARNING - possible iterative failure at 491 with rmsd of 2.00661
WARNING - possible iterative failure at 600 with rmsd of 2.00064
WARNING - possible iterative failure at 703 with rmsd of 2.00268
WARNING - possible iterative failure at 821 with rmsd of 2.00095
*** There were 6 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.3004
Aligned rmsd = 0.994363
[0.824525 -0.00479267 -0.565805 0.289809;0.00447157 0.999988 -0.00195419 0.0630819;0.565808 -0.000918763 0.824536 -0.153339;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 4 with rmsd of 2.00199
WARNING - possible iterative failure at 148 with rmsd of 2.00124
WARNING - possible iterative failure at 310 with rmsd of 2.00198
WARNING - possible iterative failure at 471 with rmsd of 2.00076
WARNING - possible iterative failure at 489 with rmsd of 2.00116
WARNING - possible iterative failure at 834 with rmsd of 2.00054
WARNING - possible iterative failure at 850 with rmsd of 2.00002
WARNING - possible iterative failure at 898 with rmsd of 2.00153
*** There were 8 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #3 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.2435
Aligned rmsd = 0.953744
[0.826399 -0.00892513 -0.563014 -0.0653625;-0.0208073 0.998707 -0.0463732 2.29056;0.562701 0.0500376 0.825145 -3.24716;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 370 with rmsd of 2.00008
*** There were 1 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.1667
Aligned rmsd = 0.9819
[0.840057 0.0332238 -0.541481 0.111471;-0.0382095 0.999268 0.00203399 -1.45935;0.541152 0.018981 0.840711 -1.53685;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 228 with rmsd of 2.01266
WARNING - possible iterative failure at 774 with rmsd of 2.0063
WARNING - possible iterative failure at 842 with rmsd of 2.00158
*** There were 3 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.256
Aligned rmsd = 0.991809
[0.817903 -0.014588 -0.575171 0.278167;0.00628017 0.999845 -0.0164285 0.103457;0.575322 0.00982477 0.817868 -0.185224;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 263 with rmsd of 2.00077
WARNING - possible iterative failure at 412 with rmsd of 2.0015
WARNING - possible iterative failure at 446 with rmsd of 2.0038
WARNING - possible iterative failure at 786 with rmsd of 2.00096
WARNING - possible iterative failure at 883 with rmsd of 2.00024
*** There were 5 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************


-=-=-=-=-=-=-=- MACROCYCLE #4 -=-=-=-=-=-=-=-
**************************************************
Selected 28 atoms with selector 'name == 'CA''.
RMSD test = 1
Rotated rmsd = 36.329
Aligned rmsd = 0.943436
[0.818476 -0.017062 -0.574288 1.17034;0.0345878 0.999209 0.0196082 -2.45155;0.573499 -0.0359122 0.818419 2.70253;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 674 with rmsd of 2.00229
WARNING - possible iterative failure at 979 with rmsd of 2.00247
WARNING - possible iterative failure at 996 with rmsd of 2.00639
*** There were 3 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 25 atoms with selector 'resid == 5'.
RMSD test = 1
Rotated rmsd = 22.0823
Aligned rmsd = 0.987444
[0.828393 0.0419179 -0.558577 -0.702659;-0.0102908 0.998167 0.0596447 -0.0850405;0.560053 -0.0436611 0.827305 0.703891;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 6
WARNING - possible iterative failure at 382 with rmsd of 2.00671
WARNING - possible iterative failure at 694 with rmsd of 2.00275
*** There were 2 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
**************************************************
Selected 125 atoms with selector 'segid == 'L2''.
RMSD test = 1
Rotated rmsd = 17.4037
Aligned rmsd = 0.982243
[0.816913 0.00954453 -0.576681 -0.196444;-0.00263654 0.999914 0.0128145 0.047971;0.576754 -0.0089479 0.816869 0.377722;0 0 0 1;];
*** Single Alignment Tests ***
All tests passed (threshold = 1e-06)
Total iterations = 4
WARNING - possible iterative failure at 503 with rmsd of 2.00117
WARNING - possible iterative failure at 526 with rmsd of 2.00022
WARNING - possible iterative failure at 734 with rmsd of 2.00019
WARNING - possible iterative failure at 752 with rmsd of 2.00019
*** There were 4 possible errors detected.
All composite iterative tests passed (threshold = 1e-06)
**************************************************
